src/librustc_trans/partitioning.rs

```diff
@@ -0,0 +1,376 @@
+//! using such a highly granular partitioning would lead to very slow runtime
+//! code since it would effectively prohibit inlining and other inter-procedure
+//! optimizations. We want to avoid that as much as possible.
+//!
+//! Thus we end up with a trade-off: The bigger the codegen units, the better
+//! LLVM's optimizer can do its work, but also the smaller the compilation time
+//! reduction we get from incremental compilation.
+//!
+//! Ideally, we would create a partitioning such that there are few big codegen
+//! units with few interdependencies between them. For now though, we use the
+//! following heuristic to determine the partitioning:
+//!
+//! - There are two codegen units for every source-level module:
+//! - One for "stable", that is non-generic, code
+//! - One for more "volatile" code, i.e. monomorphized instances of functions
+//!   defined in that module
+//! - Code for monomorphized instances of functions from external crates gets
+//!   placed into every codegen unit that uses that instance.
+//!
+//! In order to see why this heuristic makes sense, let's take a look at when a
+//! codegen unit can get invalidated:
+//!
+//! 1. The most straightforward case is when the BODY of a function or global
+//! changes. Then any codegen unit containing the code for that item has to be
+//! re-compiled. Note that this includes all codegen units where the function
+//! has been inlined.
+//!
+//! 2. The next case is when the SIGNATURE of a function or global changes. In
+//! this case, all codegen units containing a REFERENCE to that item have to be
+//! re-compiled. This is a superset of case 1.
+//!
+//! 3. The final and most subtle case is when a REFERENCE to a generic function
+//! is added or removed somewhere. Even though the definition of the function
+//! might be unchanged, a new REFERENCE might introduce a new monomorphized
+//! instance of this function which has to be placed and compiled somewhere.
+//! Conversely, when removing a REFERENCE, it might have been the last one with
+//! that particular set of generic arguments and thus we have to remove it.
+//!
+//! From the above we see that just using one codegen unit per source-level
+//! module is not such a good idea, since just adding a REFERENCE to some
+//! generic item somewhere else would invalidate everything within the module
+//! containing the generic item. The heuristic above reduces this detrimental
+//! side-effect of references a little by at least not touching the non-generic
+//! code of the module.
+//!
+//! As another optimization, monomorphized functions from external crates get
+//! some special handling. Since we assume that the definition of such a
+//! function never changes, we can just instantiate such a function in every
+//! codegen unit where it is referenced -- without having to fear that doing
+//! this will cause a lot of unnecessary re-compilations. If such a reference is
```

Comment:
Why do we assume these functions never change? I am currently working on a patch, for example, that traces changes across crates. This optimization will mean more recompilation than would otherwise be needed, I think. (That is, it causes maximal damage to downstream crates, rather than minimal, if random damange.)

Nonetheless, this is probably a good policy. Otherwise we biased towards one CGU -- which gets definitions local. Hmm, but another option would be to make groupings also for modules in extern crates -- or just one CGU per extern crate -- so that only code from that crate is affected.

Also, you don't write about `#[inline]`, but I think it should feed into the heuristics here.